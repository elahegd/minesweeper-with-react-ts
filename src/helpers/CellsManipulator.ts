// Increment neighbour items for cell with coords

import { CellState, Cell, Coords, Field } from "./Field";

export const getNeigboursItems = ([y, x]: Coords): Record<string, [number, number]> => ({
    top: [y-1, x],
    topRight: [y-1, x+1],
    right: [y, x+1],
    rightBottom: [y+1, x+1],
    bottom: [y+1, x],
    left: [y, x-1],
    leftTop: [y-1, x-1],
    bottomLeft: [y+1, x-1]
});

export const checkItemInField = ([y, x]: Coords, { length }: Field): boolean => {
    return y >= 0 && x >= 0 && length - y > 0 && length - x > 0;
}
    

export const incrementNeibours = (coords: Coords, field: Field): Field => {
    const items = getNeigboursItems(coords);
    // console.log("hhh", items, Object.values(items), field);
    for (const [y, x] of Object.values(items)) {
        if (checkItemInField([y, x], field)) {
          const cell = field[y][x];
          // wnat to show the digit   
          if (cell < 8) {
            // increment the cell because we want to get rid of empty which is 0 and show the digit
            field[y][x] = (cell + 1) as Cell;
          }
        }
      }
    return field;
}
