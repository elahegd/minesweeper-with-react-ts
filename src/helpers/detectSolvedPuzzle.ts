import { Field, CellState } from "./Field";

export const detectSolvedPuzzle = (
    playerField: Field, 
    gameField:Field): [boolean, number] => {
        const { empty, flag, weakFlag, bomb, hidden } = CellState;
        let isFieldHaveHiddenCell: boolean = false;
        let bombCounter: number = 0;
        let detectedBomb: number = 0;
        let flagCounter: number = 0;

        for(const y of gameField.keys()) {
            for(const x of gameField[y].keys()) {
                const playerCell = playerField[y][x];
                const gameCell = gameField[y][x];
                const flagged = [flag, weakFlag].includes(playerCell);
                
                if(playerCell === hidden) {
                    isFieldHaveHiddenCell = true 
                }
                
                if(flagged) {
                    
                    flagCounter ++;
                }
                if(gameCell === bomb) {
                    bombCounter++;

                    if(flagged) {
                        detectedBomb++;
                    }
                }
            }
        }
        
        const isPuzzleSolved = 
            bombCounter === detectedBomb &&
            flagCounter === bombCounter &&
            !isFieldHaveHiddenCell;

        return [isPuzzleSolved, flagCounter];
}