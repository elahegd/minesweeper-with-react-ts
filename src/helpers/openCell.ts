import { getNeigboursItems, checkItemInField } from "./CellsManipulator";
import { copyField } from "./copyField";
import { detectSolvedPuzzle } from "./detectSolvedPuzzle";
import { CellState, Coords, Field } from "./Field";

export const openCell = (coords: Coords, playerField: Field, gameField: Field): [Field, boolean] => {
    return openCellRecursively(coords, copyField(playerField), gameField);
}

export const openCellRecursively = (coords: Coords, playerField: Field, gameField: Field): [Field, boolean] => {
    const { bomb, hidden, empty, flag, weakFlag } = CellState;

    const [y, x] = coords;
    const gameCell = gameField[y][x];
    const playerCell = playerField[y][x];

    if(gameCell === bomb) {
        throw new Error("Game Over");
    }

    if (flag === playerCell) {
        return [playerField, false];
    }

    playerField[y][x] = gameCell;
    if(gameCell === empty && [hidden, weakFlag].includes(playerCell)) {
        const neighbour = getNeigboursItems(coords);
        
        for (const [y, x] of Object.values(neighbour)) {
            if(checkItemInField([y, x], gameField)) {
                [playerField] = openCellRecursively([y, x], playerField, gameField);
            }
        }
    }
    
    const [isSolved] = detectSolvedPuzzle(playerField, gameField)
    
    return [playerField, isSolved];
}