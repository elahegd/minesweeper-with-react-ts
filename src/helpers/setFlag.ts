import { copyField } from "./copyField";
import { detectSolvedPuzzle } from "./detectSolvedPuzzle";
import { CellState, Coords, Field } from "./Field";

const { flag, weakFlag, hidden } = CellState;

export const setFlag = (
    coords: Coords, 
    playerField: Field, 
    gameField: Field,
    prevFlagCounter: number,
    bombsCounter: number
    ): [Field, boolean, number] => { 
        const [y,x] = coords;
        const newPlayerField = copyField(playerField)

        const cell = newPlayerField[y][x];

        const { flag, hidden, weakFlag } = CellState;

        switch(cell) {
            case flag:
                newPlayerField[y][x] = weakFlag;
                break;
            case hidden:
                if(prevFlagCounter < bombsCounter) {
                    newPlayerField[y][x] = flag;
                }
                break;
            case weakFlag:
                newPlayerField[y][x] = hidden;
                break;
        }

        const [isSolved, flagCounter] = detectSolvedPuzzle(newPlayerField, gameField);
        return [newPlayerField, isSolved, flagCounter];
}