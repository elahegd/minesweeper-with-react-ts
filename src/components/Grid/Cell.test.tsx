import React from 'react';
import {render, screen, fireEvent, createEvent} from "@testing-library/react";

import { CellState, Coords } from '../../helpers/Field';
import Cell, { checkCellIsActive, ClosedFrame } from "./Cell";

describe('Cell component check', () => {
    const coords: Coords = [1, 1];
    it("Should render elements of cell with props", () => {
        const props = {
            onClick: jest.fn(), 
            onContextMenu: jest.fn(),
            coords: coords,            
        }
        render(<Cell {...props}  children= {0} />);
    });

    it("Closed frame renders correctly", () => {
        const { asFragment } = render(<ClosedFrame mouseDown={true} />);
        expect(asFragment()).toMatchSnapshot();
    })

    for(let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
        it("check prevent default context menu for every type of cell", () => {
            const props = {
                coords,
                onClick: jest.fn(),
                onContextMenu: jest.fn(),
            }
            render(<Cell {...props} children= {cell} />);

            //should pass this data-testid to the element in the component as well
            const cellComponent = screen.getByTestId(`${coords}`);
            const contextMenuEvent = createEvent.contextMenu(cellComponent);
            fireEvent(cellComponent, contextMenuEvent);
            expect(contextMenuEvent.defaultPrevented).toBe(true);
        });

        it("onClick and onContextMenu handler shhould be called for active cells", () => {
            const props = {
                coords,
                onClick: jest.fn(),
                onContextMenu: jest.fn(),
                'data-testid':`${coords}`
            }
            render(<Cell {...props} children= {cell} />);

            const cellComponent = screen.getByTestId(`${coords}`);
            fireEvent.contextMenu(cellComponent);
            fireEvent.click(cellComponent);

            if(checkCellIsActive(cell)){
                expect(props.onClick).toBeCalled();
                expect(props.onContextMenu).toBeCalled();
            } else {
                expect(props.onClick).not.toBeCalled();
                expect(props.onContextMenu).not.toBeCalled();
            }
        }); 
    }

})