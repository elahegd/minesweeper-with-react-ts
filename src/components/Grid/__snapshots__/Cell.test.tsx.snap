// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component check Closed frame renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;
